scio-core/src/main/scala/com/spotify/scio/util/BloomFilter.scala

```diff
@@ -0,0 +1,576 @@
+   * This may be faster if you don't care about evaluating
+   * the false positive probability
+   */
+  def maybeContains(item: A): Boolean
+
+  // Estimates the cardinality of the set of elements that have been
+  // inserted into the Bloom Filter.
+  def size: Approximate[Long]
+
+  def toBitSet: util.BitSet
+
+  def copy: MutableBF[A]
+
+  /**
+   * Compute the Hamming distance between the two Bloom filters
+   * `a` and `b`. The distance is defined as the number of bits that
+   * need to change to in order to transform one filter into the other.
+   * This is computed using XOR but it doesn't mutate any BloomFilters
+   */
+  def hammingDistance(that: MutableBF[A]): Int =
+    (this, that) match {
+      // Comparing with empty filter should give number
+      // of bits in other set
+      case (x: MutableBFZero[A], y: MutableBFZero[A]) => 0
+      case (x: MutableBFZero[A], y: MutableBF[A])     => y.numBits
+      case (x: MutableBF[A], y: MutableBFZero[A])     => x.numBits
+
+      // Otherwise compare as bit sets
+      case (_, _) =>
+        // hammingDistance should not mutate BloomFilter
+        val thisCopy = this.toBitSet.clone().asInstanceOf[util.BitSet]
+        thisCopy.xor(that.toBitSet)
+        thisCopy.cardinality()
+    }
+
+}
+
+/**
+ * Empty bloom filter.
+ */
+private[scio] final case class MutableBFZero[A](hashes: KirMit32Hash[A]) extends MutableBF[A] {
+
+  lazy val toBitSet: util.BitSet = new util.BitSet()
+
+  lazy val numHashes: Int = hashes.numHashes
+
+  lazy val width: Int = hashes.width
+
+  lazy val numBits: Int = 0
+
+  // scalastyle:off method.name
+  def ++=(other: MutableBF[A]): MutableBF[A] = other
+
+  def +=(other: A): MutableBF[A] = MutableBFInstance[A](hashes, other)
+  // scalastyle:on method.name
+
+  def checkAndAdd(other: A): (MutableBF[A], ApproximateBoolean) =
+    (this += other, ApproximateBoolean.exactFalse)
+
+  override def contains(item: A): ApproximateBoolean = ApproximateBoolean.exactFalse
+
+  def maybeContains(item: A): Boolean = false
+
+  lazy val size: Approximate[Long] = Approximate.exact[Long](0)
+
+  def copy: MutableBF[A] = MutableBFZero(hashes)
+}
+
+/**
+ * Mutable Bloom filter with multiple values
+ */
+final case class MutableBFInstance[A](hashes: KirMit32Hash[A], bits: util.BitSet)
+    extends MutableBF[A] {
+
+  lazy val numHashes: Int = hashes.numHashes
+
+  /**
+   * The number of bits set to true
+   */
+  def numBits: Int = bits.cardinality()
+
+  lazy val width: Int = hashes.width
+
+  def toBitSet: util.BitSet = bits
+
+  // scalastyle:off method.name
+  def ++=(other: MutableBF[A]): MutableBF[A] = {
+    require(this.width == other.width)
+    require(this.numHashes == other.numHashes)
+
+    other match {
+      case MutableBFZero(_)                         => this
+      case MutableSparseBFInstance(_, otherSetBits) =>
+        // This is MutableBFInstance, hence not sparse, so don't convert output to sparse.
+        val it = otherSetBits.flatten.iterator
```

Comment:
invert this to avoid extra copy `.iterator.flatten`

Reply:
Thanks, this is hard to notice. :)